File: src-tauri/src/debug_tree.rs
/* Placeholder ParserInfo structures for state management */
#[derive(Clone, Debug, PartialEq, serde::Serialize)]
pub struct DebugTree {
    input: String,
    root: DebugNode,
}

impl DebugTree {
    pub fn new(input: String, root: DebugNode) -> Self {
        DebugTree { input, root }
    }

    pub fn get_root(&self) -> &DebugNode {
        &self.root
    }

    pub fn get_input(&self) -> &String {
        &self.input
    }
}

/* Defines tree structure used in backend that will be passed to frontend */
#[derive(Debug, Clone, PartialEq, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct DebugNode {
    pub node_id: u32,          /* The unique number of this node */
    pub name: String,          /*The internal (default) or user-defined name of the parser */
    pub internal: String,      /*The internal name of the parser */
    pub success: bool,         /* Whether the parser was successful */
    pub child_id: Option<u32>, /* The unique child number of this node */
    pub input: String,         /* The input string passed to the parser */
    #[serde(skip_serializing)]
    pub children: Vec<DebugNode>, /* The children of this node */
    pub is_leaf: bool,
}

impl DebugNode {
    pub fn new(
        node_id: u32,
        name: String,
        internal: String,
        success: bool,
        child_id: Option<u32>,
        input: String,
        children: Vec<DebugNode>,
    ) -> Self {
        let is_leaf: bool = children.is_empty();
        DebugNode {
            node_id,
            name,
            internal,
            success,
            child_id,
            input,
            children,
            is_leaf,
        }
    }
}

File: src-laminar/src/main/scala/model/DebugTree.scala
package model

import upickle.default.ReadWriter

/**
  * Case class used to represent tauri's debug tree automatically using upickle.
  *
  * @param input the input string the tree node has to parse
  * @param root the debug tree root node
  */
case class DebugTree(input: String, root: DebugNode) derives ReadWriter

/**
  * Case class used to represent tauri's debug tree automatically using upickle.
  *
  * @param nodeId an unique identifier number for the tree node
  * @param name user defined name of the debug tree node
  * @param internal internal name of the debug tree node
  * @param success whether the input was successfully consumed
  * @param childId an unique identifier number for the tree node
  * @param input the input string the tree node has to parse
  * @param children list of children nodes
  * @param isLeaf if a node is a leaf
  */
case class DebugNode(nodeId: Int, name: String, internal: String, success: Boolean,
    childId: Int, input: String, children: List[DebugNode], isLeaf: Boolean) derives ReadWriter

object DebugTree {
    lazy final val Sample: DebugTree = DebugTree(
        "hi world!",
        DebugNode(0, "~>", "~>", true, 0, "hi world!", List[DebugNode](
            DebugNode(1, "~>", "~>", true, 1, "hi", List[DebugNode](
                DebugNode(2, "'h'", "'h'", true, 1, "h", List[DebugNode](), true),
                DebugNode(3, "|", "|", true, 2, "i", List[DebugNode](
                    DebugNode(4, "\"ello\"", "\"ello\"", false, 0, "", List[DebugNode](), true),
                    DebugNode(5, "\"i\"", "\"i\"", true, 1, "i", List[DebugNode](), true)
                ), false)
            ), false),
            DebugNode(6, "\" world!\"", "\" world!\"", true, 2, "world!", List[DebugNode](), true)
        ), false)
    )
}
